<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Document</title>
</head>
<body>
    

    <script>
/*
        function cuadradoCallback(value,callback) {
            setTimeout(() => {
                callback(value,value*value);
            }, 0 | Math.random()*1000);
        }

        cuadradoCallback(0,(value, result) => {
            console.log("Inicia Callback");
            console.log(`Callback:${value},${result}`);
            cuadradoCallback(1,(value, result) => {
            console.log(`Callback:${value},${result}`);
            cuadradoCallback(2,(value, result) => {
            console.log(`Callback:${value},${result}`);
            cuadradoCallback(3,(value, result) => {
            console.log(`Callback:${value},${result}`);
            cuadradoCallback(4,(value, result) => {
            console.log(`Callback:${value},${result}`);
            cuadradoCallback(5,(value, result) => {
            console.log(`Callback:${value},${result}`);
            console.log(`Fin mi pana`);
            
        })
        })
        })
        })
        })
        })


function cuadradoPromise(value) {
    if (typeof value  !== "number") {
    return Promise.reject(`Erroe el valor"${value}"ingresado no es un número`);
        }
    return new Promise((resolve, reject) => {
            setTimeout(() => {
                resolve({
                    value:value,
                    result:value*value
                });
            }, 0 | Math.random()*1000);
        });
        }

        cuadradoPromise(0)
        .then(obj=>{
            console.log(obj);
            console.log("Inicio promise");
            console.log(`Promise: ${obj.value},${obj.result}`);
            return cuadradoPromise(1);
        })
        .then(obj =>{
        console.log(`Promise:${obj.value},${obj.result}`);
        return cuadradoPromise(2);
        })
        .then(obj =>{
        console.log(`Promise:${obj.value},${obj.result}`);
        return cuadradoPromise(3);
        })
        .then(obj =>{
        console.log(`Promise:${obj.value},${obj.result}`);
        return cuadradoPromise(4);
        })
        .then(obj =>{
        console.log(`Promise:${obj.value},${obj.result}`);
        return cuadradoPromise();
        })
        .then(obj =>{
        console.log(`Promise:${obj.value},${obj.result}`);
        console.log("Fin Promise")
        })
        .catch(err => console.error(err));


function cuadradoPromise(value) {
    if (typeof value  !== "number") {
    return Promise.reject(`Erroe el valor"${value}"ingresado no es un número`);
        }
    return new Promise((resolve, reject) => {
            setTimeout(() => {
                resolve({
                    value:value,
                    result:value*value
                });
            }, 0 | Math.random()*1000);
        });
        }

        async function funcionAsyncronaDeclarada(){
            try {
                console.log('Inicio Async Functions');

                let obj = await cuadradoPromise(0);
                console.log(`Async Fuñnction: ${obj.value},${obj.result}`)

                obj =await cuadradoPromise(1);
                console.log(`Async Fuñnction: ${obj.value},${obj.result}`)
                obj =await cuadradoPromise(2);
                console.log(`Async Fuñnction: ${obj.value},${obj.result}`)
                obj =await cuadradoPromise(3);
                console.log(`Async Fuñnction: ${obj.value},${obj.result}`)
                obj =await cuadradoPromise(4);
                console.log(`Async Fuñnction: ${obj.value},${obj.result}`)
                obj =await cuadradoPromise("s");
                console.log(`Async Fuñnction: ${obj.value},${obj.result}`)
                
                console.log('fin de esto ome');
            }catch(err){
                console.error(err);
            }
        }

        funcionAsyncronaDeclarada();

        const funcionAsincronaExpresada =async ()=>{
            try {
                console.log('Inicio Async Functions');

                let obj = await cuadradoPromise(6);
                console.log(`Async Fuñnction: ${obj.value},${obj.result}`)

                obj =await cuadradoPromise(7);
                console.log(`Async Fuñnction: ${obj.value},${obj.result}`)
                obj =await cuadradoPromise(8);
                console.log(`Async Fuñnction: ${obj.value},${obj.result}`)
                obj =await cuadradoPromise(9);
                console.log(`Async Fuñnction: ${obj.value},${obj.result}`)
                obj =await cuadradoPromise(10);
                console.log(`Async Fuñnction: ${obj.value},${obj.result}`)
              
                console.log('fin de esto ome');
            }catch(err){
                console.error(err);
            }
        }
        funcionAsincronaExpresada();
        */










        
    </script>








</body>
</html>